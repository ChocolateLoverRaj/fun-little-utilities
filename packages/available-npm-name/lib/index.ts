export * from './constants'
export * from './fmt'
export * from './main'
export * from './parse-input'
export * from './types'
export * from './unit'
