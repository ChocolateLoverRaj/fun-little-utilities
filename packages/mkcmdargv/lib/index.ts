import { partition } from 'ramda'
import { splitIterable } from 'split-iterable'

type Value = string | number | boolean

type Options = {
  readonly [name in string]: Value
}

export interface Param {
  readonly args?: readonly string[]
  readonly options?: Options
  readonly flags?: readonly string[]
}

function parseSingleDash ([key, value]: [string, Value]) {
  const optName = '-' + key
  if (value === true) return [optName]
  if (value === false) return []
  return [optName, String(value)]
}

function parseDoubleDash ([key, value]: [string, Value]) {
  const optName = '--' + key
  if (value === true) return [optName]
  if (value === false) return []
  return [`${optName}=${value}`]
}

export function * iterateCommandArguments (param: Param) {
  const { args = [], options = {}, flags = [] } = param
  const [beforeDoubleDashes, ...afterDoubleDashes] = splitIterable(args, x => x === '--')
  const [beforeDashes, afterDashes] = partition(x => x[0] !== '-', beforeDoubleDashes)
  const optionsPairs = Object.entries(options)
  const [singleDashOptions, doubleDashOptions] = partition(([key]) => key.length === 1, optionsPairs)
  const [singleDashFlags, doubleDashFlags] = partition(x => x.length === 1, flags)

  yield * beforeDashes

  if (singleDashFlags.length) yield '-' + singleDashFlags.join('')

  for (const entry of singleDashOptions) {
    yield * parseSingleDash(entry)
  }

  yield * doubleDashFlags.map(x => '--' + x)

  for (const entry of doubleDashOptions) {
    yield * parseDoubleDash(entry)
  }

  if (!afterDashes.length && !afterDoubleDashes.length) return

  yield '--'
  yield * afterDashes

  for (const chunk of afterDoubleDashes) {
    yield '--'
    yield * chunk
  }
}

export function createCommandArguments (param: Param) {
  return Array.from(iterateCommandArguments(param))
}

export async function help () {
  const pathPromise = import('path')
  const fsxPromise = import('fs-extra')
  const path = await pathPromise
  const helpFilePath = path.resolve(__dirname, '../data/help.txt')
  const { readFile } = await fsxPromise
  return readFile(helpFilePath, { encoding: 'utf8' })
}
