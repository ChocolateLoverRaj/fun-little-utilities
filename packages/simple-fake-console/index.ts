export * from './action'
export * from './console'
export * from './string'
