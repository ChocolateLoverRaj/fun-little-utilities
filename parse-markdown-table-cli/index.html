<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>parse-markdown-table-cli — Reference</title>
	<meta name="description" content="">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="assets/css/main.css">
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="assets/js/search.js" data-base=".">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="index.html" class="title">parse-markdown-table-cli — Reference</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="globals.html">Globals</a>
				</li>
			</ul>
			<h1>parse-markdown-table-cli — Reference</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<div class="tsd-panel tsd-typography">
				<a href="#parse-markdown-table-cli" id="parse-markdown-table-cli" style="color: inherit; text-decoration: none;">
					<h1>Parse Markdown Table CLI</h1>
				</a>
				<p>Convert a markdown table from text to JSON objects</p>
				<a href="#usage" id="usage" style="color: inherit; text-decoration: none;">
					<h2>Usage</h2>
				</a>
				<a href="#code--helpcode" id="code--helpcode" style="color: inherit; text-decoration: none;">
					<h3><code>--help</code></h3>
				</a>
				<pre><code><span class="hljs-keyword">input</span>-stream | <span class="hljs-keyword">parse</span>-markdown-<span class="hljs-keyword">table</span> [options]

Options:
  --<span class="hljs-keyword">version</span>     Show <span class="hljs-keyword">version</span> number                                                     [boolean]
  --<span class="hljs-keyword">format</span>      Structure of output          [choices: <span class="hljs-string">"list"</span>, <span class="hljs-string">"dict"</span>, <span class="hljs-string">"jsonl"</span>] [default: <span class="hljs-string">"dict"</span>]
  --indentType  <span class="hljs-keyword">Type</span> of indentation          [choices: <span class="hljs-string">"tab"</span>, <span class="hljs-string">"space"</span>, <span class="hljs-string">"none"</span>] [default: <span class="hljs-string">"space"</span>]
  --indentSize  Size of indentation
                (only apply <span class="hljs-keyword">if</span> --indentType=space)                          [number] [default: 2]
  --<span class="hljs-keyword">help</span>        Show <span class="hljs-keyword">help</span>                                                               [boolean]

Examples:
  <span class="hljs-keyword">parse</span>-markdown-<span class="hljs-keyword">table</span> &lt; <span class="hljs-keyword">table</span>.md  <span class="hljs-keyword">Print</span> JSON representation of <span class="hljs-keyword">table</span> inside <span class="hljs-keyword">table</span>.md
  <span class="hljs-keyword">parse</span>-markdown-<span class="hljs-keyword">table</span>             <span class="hljs-keyword">Read</span> a markdown <span class="hljs-keyword">table</span> from stdin and <span class="hljs-keyword">parse</span> it</code></pre>
				<a href="#examples" id="examples" style="color: inherit; text-decoration: none;">
					<h3>Examples</h3>
				</a>
				<a href="#code--format-dictcode" id="code--format-dictcode" style="color: inherit; text-decoration: none;">
					<h4><code>--format dict</code></h4>
				</a>
				<pre><code class="language-sh"><span class="hljs-built_in">echo</span> <span class="hljs-string">'
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
'</span> | parse-markdown-table --format dict</code></pre>
				<p><strong>Output:</strong></p>
				<pre><code class="language-json">[
  {
    <span class="hljs-attr">"id"</span>: <span class="hljs-string">"1"</span>,
    <span class="hljs-attr">"name"</span>: <span class="hljs-string">"John Doe"</span>,
    <span class="hljs-attr">"email"</span>: <span class="hljs-string">"john-doe@gmail.com"</span>
  },
  {
    <span class="hljs-attr">"id"</span>: <span class="hljs-string">"2"</span>,
    <span class="hljs-attr">"name"</span>: <span class="hljs-string">"Peter Smith"</span>,
    <span class="hljs-attr">"email"</span>: <span class="hljs-string">"petersmith22@outlook.com"</span>
  },
  {
    <span class="hljs-attr">"id"</span>: <span class="hljs-string">"3"</span>,
    <span class="hljs-attr">"name"</span>: <span class="hljs-string">"Julia Jones"</span>,
    <span class="hljs-attr">"email"</span>: <span class="hljs-string">"jjones778@gmail.com"</span>
  }
]</code></pre>
				<a href="#code--format-listcode" id="code--format-listcode" style="color: inherit; text-decoration: none;">
					<h4><code>--format list</code></h4>
				</a>
				<pre><code class="language-sh"><span class="hljs-built_in">echo</span> <span class="hljs-string">'
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
'</span> | parse-markdown-table --format list</code></pre>
				<p><strong>Output:</strong></p>
				<pre><code class="language-json">{
  <span class="hljs-attr">"headers"</span>: [
    <span class="hljs-string">"id"</span>,
    <span class="hljs-string">"name"</span>,
    <span class="hljs-string">"email"</span>
  ],
  <span class="hljs-attr">"rows"</span>: [
    [
      <span class="hljs-string">"1"</span>,
      <span class="hljs-string">"John Doe"</span>,
      <span class="hljs-string">"john-doe@gmail.com"</span>
    ],
    [
      <span class="hljs-string">"2"</span>,
      <span class="hljs-string">"Peter Smith"</span>,
      <span class="hljs-string">"petersmith22@outlook.com"</span>
    ],
    [
      <span class="hljs-string">"3"</span>,
      <span class="hljs-string">"Julia Jones"</span>,
      <span class="hljs-string">"jjones778@gmail.com"</span>
    ]
  ]
}</code></pre>
				<a href="#code--format-jsonlcode" id="code--format-jsonlcode" style="color: inherit; text-decoration: none;">
					<h4><code>--format jsonl</code></h4>
				</a>
				<pre><code class="language-sh"><span class="hljs-built_in">echo</span> <span class="hljs-string">'
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
'</span> | parse-markdown-table --format jsonl</code></pre>
				<pre><code class="language-json">{<span class="hljs-attr">"id"</span>:<span class="hljs-string">"1"</span>,<span class="hljs-attr">"name"</span>:<span class="hljs-string">"John Doe"</span>,<span class="hljs-attr">"email"</span>:<span class="hljs-string">"john-doe@gmail.com"</span>}
{<span class="hljs-attr">"id"</span>:<span class="hljs-string">"2"</span>,<span class="hljs-attr">"name"</span>:<span class="hljs-string">"Peter Smith"</span>,<span class="hljs-attr">"email"</span>:<span class="hljs-string">"petersmith22@outlook.com"</span>}
{<span class="hljs-attr">"id"</span>:<span class="hljs-string">"3"</span>,<span class="hljs-attr">"name"</span>:<span class="hljs-string">"Julia Jones"</span>,<span class="hljs-attr">"email"</span>:<span class="hljs-string">"jjones778@gmail.com"</span>}</code></pre>
				<a href="#license" id="license" style="color: inherit; text-decoration: none;">
					<h2>License</h2>
				</a>
				<p><a href="https://git.io/fxKXN">MIT</a> © <a href="https://github.com/KSXGitHub">Hoàng Văn Khải</a></p>
			</div>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class="globals  ">
						<a href="globals.html"><em>Globals</em></a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
					<li class=" tsd-kind-enum">
						<a href="enums/format.html" class="tsd-kind-icon">Format</a>
					</li>
					<li class=" tsd-kind-enum">
						<a href="enums/indenttype.html" class="tsd-kind-icon">Indent<wbr>Type</a>
					</li>
					<li class=" tsd-kind-enum">
						<a href="enums/status.html" class="tsd-kind-icon">Status</a>
					</li>
					<li class=" tsd-kind-interface">
						<a href="interfaces/mainparam.html" class="tsd-kind-icon">Main<wbr>Param</a>
					</li>
					<li class=" tsd-kind-interface">
						<a href="interfaces/readablestream.html" class="tsd-kind-icon">Readable<wbr>Stream</a>
					</li>
					<li class=" tsd-kind-function tsd-has-type-parameter">
						<a href="globals.html#getasyncarray" class="tsd-kind-icon">get<wbr>Async<wbr>Array</a>
					</li>
					<li class=" tsd-kind-function">
						<a href="globals.html#getindentargument" class="tsd-kind-icon">get<wbr>Indent<wbr>Argument</a>
					</li>
					<li class=" tsd-kind-function">
						<a href="globals.html#main" class="tsd-kind-icon">main</a>
					</li>
					<li class=" tsd-kind-function">
						<a href="globals.html#output" class="tsd-kind-icon">output</a>
					</li>
					<li class=" tsd-kind-function">
						<a href="globals.html#readstream" class="tsd-kind-icon">read<wbr>Stream</a>
					</li>
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
			<ul class="tsd-legend">
				<li class="tsd-kind-property tsd-parent-kind-interface"><span class="tsd-kind-icon">Property</span></li>
			</ul>
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="assets/js/main.js"></script>
<script>if (location.protocol == 'file:') document.write('<script src="assets/js/search.js"><' + '/script>');</script>
</body>
</html>