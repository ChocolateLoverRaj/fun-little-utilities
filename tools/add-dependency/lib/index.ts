import path from 'path'
import { prompt } from 'inquirer'
import {
  Manifest,
  ManifestType,
  PackageManifest,
  TestManifest,
  ToolManifest,
  ManifestItem,
  ManifestList,
  NonRootManifestTypes,
  loadPackageList,
  loadTestList,
  loadToolList
} from '@tools/utils'

async function promptManifestType (): Promise<NonRootManifestTypes.Union> {
  const { type } = await prompt({
    name: 'type',
    message: 'Dependant container',
    type: 'list',
    choices: NonRootManifestTypes
      .map(value => ({ name: value.toLowerCase(), value }))
  })

  return type
}

interface GetVersionRequirementParam {
  readonly item: ManifestItem<PackageManifest>
  readonly manifest: PackageManifest
}

async function addProdDeps<Obj extends Manifest> (
  target: ManifestItem<Obj>,
  packages: ManifestList<PackageManifest>,
  getVersionRequirement: (param: GetVersionRequirementParam) => string
) {
  if (!packages.size) return

  const [
    manifest,
    dependencies
  ] = await Promise.all([
    target.readManifest(),
    Promise.all(packages.items().map(async item => ({
      item,
      manifest: await item.readManifest()
    })))
  ])

  if (!manifest.dependencies) {
    // @ts-ignore
    manifest.dependencies = {}
  }

  for (const param of dependencies) {
    const version = getVersionRequirement(param)
    // @ts-ignore
    manifest.dependencies[param.manifest.name] = version
  }

  await target.writeManifest(manifest)
}

export async function handlePackage () {
  const list = await loadPackageList()
  const target = await list.promptItem('Choose a package')
  const dependencies = await list.promptItemList('Choose dependencies')
  await addProdDeps(target, dependencies, param => '^' + param.manifest.version)
}

async function handleTestOrTool<Manifest extends TestManifest | ToolManifest> (
  loadTargetList: () => Promise<ManifestList<Manifest>>
) {
  const [
    targetList,
    packageList
  ] = await Promise.all([
    loadTargetList(),
    loadPackageList()
  ])

  const target = await targetList.promptItem('Choose a folder')
  const dependencies = await packageList.promptItemList('Choose packages')

  await addProdDeps(
    target,
    dependencies,
    param => path.relative(target.folder, param.item.folder)
  )
}

export async function handleTest () {
  await handleTestOrTool(loadTestList)
}

export async function handleTool () {
  await handleTestOrTool(loadToolList)
}

export async function main () {
  switch (await promptManifestType()) {
    case ManifestType.Package:
      return handlePackage()
    case ManifestType.Test:
      return handleTest()
    case ManifestType.Tool:
      return handleTool()
  }
}

export default main
