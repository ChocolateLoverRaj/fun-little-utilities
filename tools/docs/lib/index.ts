import path from 'path'
import process from 'process'
import { ensureDir, pathExists } from 'fs-extra'
import { Application } from 'typedoc'
import places from '@tools/places'
import { loadPackageList } from '@tools/utils'

export async function main () {
  const failures = []

  await ensureDir(places.docs)

  const list = await loadPackageList()
  for (const item of list.items()) {
    console.info('docs>', item.name)

    const outputDir = path.join(places.docs, item.name)

    const readme = path.join(item.folder, 'README.md')
    const readmeObject = await pathExists(readme) ? { readme } : null

    const app = new Application({
      target: 'esnext',
      module: 'commonjs',
      experimentalDecorators: true,
      logger: 'none',
      exclude: ['**/node_modules', '**/.git'],
      ...readmeObject
    })

    const project = app.convert(app.expandInputFiles([item.folder]))

    if (!project) {
      failures.push(item)
      continue
    }

    app.generateDocs(project, outputDir)
  }

  if (failures.length) {
    console.error()
    console.error('[ERROR]: Failed to generate docs for:')

    for (const item of failures) {
      console.error(' -', item.name)
    }

    throw process.exit(1)
  }
}

export default main
