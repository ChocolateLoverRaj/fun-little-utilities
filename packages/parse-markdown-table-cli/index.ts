import { MarkdownCellTable, MarkdownObjectTable, ListItem } from 'parse-markdown-table'
import { Console } from 'simple-fake-console'

export interface StdinReader {
  (): Promise<string>
}

export const enum Format {
  List = 'list',
  Dict = 'dict'
}

export const enum IndentType {
  Tab = 'tab',
  Space = 'space',
  None = 'none'
}

export interface MainParam {
  readonly console: Console
  readonly getStdIn: StdinReader
  readonly format: Format
  readonly indentType: IndentType
  readonly indentSize: number
}

export const enum Status {
  Success = 0,
  Failure = 1
}

export async function main (options: MainParam): Promise<Status> {
  const indent = getIndentArgument(options.indentType, options.indentSize)
  const text = await options.getStdIn()
  const object = getOutputObject(text, options.format)
  const json = JSON.stringify(object, undefined, indent)
  options.console.info(json)
  return Status.Success
}

export function getIndentArgument (type: IndentType, size: number): '\t' | number | undefined {
  switch (type) {
    case IndentType.Tab:
      return '\t'
    case IndentType.Space:
      return size
    case IndentType.None:
      return undefined
  }
}

type OutputObject = ListItem<string, string>[] | MarkdownCellTable

export function getOutputObject (text: string, type: Format): OutputObject {
  switch (type) {
    case Format.Dict:
      return Array.from(new MarkdownObjectTable(text))
    case Format.List:
      return new MarkdownCellTable(text)
  }
}
