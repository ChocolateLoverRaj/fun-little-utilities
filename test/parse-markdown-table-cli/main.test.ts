import { ConsoleInstance, ActionType, getString } from 'simple-fake-console'
import { MainParam, Format, IndentType, Status, main } from 'parse-markdown-table-cli'
import { getAsyncArray } from './lib/async-array'

const STDIN_TEXT = `
  | id | name        | email                    |
  |----|-------------|--------------------------|
  |  1 | John Doe    | john-doe@gmail.com       |
  |  2 | Peter Smith | petersmith22@outlook.com |
  |  3 | Julia Jones | jjones778@gmail.com      |
`

class InputStream implements AsyncIterable<string> {
  async * [Symbol.asyncIterator] () {
    let length = 0
    let count = length
    let chunk = ''
    for (const char of STDIN_TEXT) {
      if (!count) {
        yield chunk
        chunk = ''
        count = length
        length += 1
      } else {
        count -= 1
      }
      chunk += char
    }
    yield chunk
  }
}

const DICT_OUTPUT = [
  {
    id: '1',
    name: 'John Doe',
    email: 'john-doe@gmail.com'
  },
  {
    id: '2',
    name: 'Peter Smith',
    email: 'petersmith22@outlook.com'
  },
  {
    id: '3',
    name: 'Julia Jones',
    email: 'jjones778@gmail.com'
  }
]

const LIST_OUTPUT = {
  headers: ['id', 'name', 'email'],
  rows: [
    ['1', 'John Doe', 'john-doe@gmail.com'],
    ['2', 'Peter Smith', 'petersmith22@outlook.com'],
    ['3', 'Julia Jones', 'jjones778@gmail.com']
  ]
}

abstract class MockedParam implements MainParam {
  public readonly console = new ConsoleInstance()
  public readonly stdin = new InputStream()
  public abstract readonly format: Format
  public abstract readonly indentType: IndentType
  public abstract readonly indentSize: number
}

async function setup (Param: new () => MockedParam) {
  const param = new Param()
  const status = await main(param)
  return { param, status }
}

describe('assertions', () => {
  it('InputParam yields multiple chunks', async () => {
    expect(await getAsyncArray(new InputStream())).toMatchSnapshot()
  })

  it('InputStream when combine is STDIN_TEXT', async () => {
    const array = await getAsyncArray(new InputStream())
    expect(array.join('')).toBe(STDIN_TEXT)
  })
})

describe('--format=dict --indentType=space --indentSize=2', () => {
  class Param extends MockedParam {
    public readonly format = Format.Dict
    public readonly indentType = IndentType.Space
    public readonly indentSize = 2
  }

  it('calls console.info', async () => {
    const { param } = await setup(Param)
    expect(param.console.getActions()).toMatchSnapshot()
  })

  it('prints expected JSON text', async () => {
    const { param } = await setup(Param)
    const text = getString({
      console: param.console,
      types: [ActionType.Info]
    })
    const object = JSON.parse(text)
    expect(object).toEqual(DICT_OUTPUT)
  })

  it('prints formatted JSON text', async () => {
    const { param } = await setup(Param)
    expect(getString({
      console: param.console,
      types: [ActionType.Info]
    })).toBe(JSON.stringify(DICT_OUTPUT, undefined, 2))
  })

  it('returns 0', async () => {
    const { status } = await setup(Param)
    expect(status).toBe(Status.Success)
  })
})

describe('--format=dict --indentType=tab', () => {
  class Param extends MockedParam {
    public readonly format = Format.Dict
    public readonly indentType = IndentType.Tab
    public readonly indentSize = NaN
  }

  it('calls console.info', async () => {
    const { param } = await setup(Param)
    expect(param.console.getActions()).toMatchSnapshot()
  })

  it('prints expected JSON text', async () => {
    const { param } = await setup(Param)
    const text = getString({
      console: param.console,
      types: [ActionType.Info]
    })
    const object = JSON.parse(text)
    expect(object).toEqual(DICT_OUTPUT)
  })

  it('prints formatted JSON text', async () => {
    const { param } = await setup(Param)
    expect(getString({
      console: param.console,
      types: [ActionType.Info]
    })).toBe(JSON.stringify(DICT_OUTPUT, undefined, '\t'))
  })

  it('returns 0', async () => {
    const { status } = await setup(Param)
    expect(status).toBe(Status.Success)
  })
})

describe('--format=dict --indentType=none', () => {
  class Param extends MockedParam {
    public readonly format = Format.Dict
    public readonly indentType = IndentType.None
    public readonly indentSize = NaN
  }

  it('calls console.info', async () => {
    const { param } = await setup(Param)
    expect(param.console.getActions()).toMatchSnapshot()
  })

  it('prints expected JSON text', async () => {
    const { param } = await setup(Param)
    const text = getString({
      console: param.console,
      types: [ActionType.Info]
    })
    const object = JSON.parse(text)
    expect(object).toEqual(DICT_OUTPUT)
  })

  it('prints minified JSON text', async () => {
    const { param } = await setup(Param)
    expect(getString({
      console: param.console,
      types: [ActionType.Info]
    })).toBe(JSON.stringify(DICT_OUTPUT))
  })

  it('returns 0', async () => {
    const { status } = await setup(Param)
    expect(status).toBe(Status.Success)
  })
})

describe('--format=list --indentType=space --indentSize=2', () => {
  class Param extends MockedParam {
    public readonly format = Format.List
    public readonly indentType = IndentType.Space
    public readonly indentSize = 2
  }

  it('calls console.info', async () => {
    const { param } = await setup(Param)
    expect(param.console.getActions()).toMatchSnapshot()
  })

  it('prints expected JSON text', async () => {
    const { param } = await setup(Param)
    const text = getString({
      console: param.console,
      types: [ActionType.Info]
    })
    const object = JSON.parse(text)
    expect(object).toEqual(LIST_OUTPUT)
  })

  it('prints formatted JSON text', async () => {
    const { param } = await setup(Param)
    expect(getString({
      console: param.console,
      types: [ActionType.Info]
    })).toBe(JSON.stringify(LIST_OUTPUT, undefined, 2))
  })

  it('returns 0', async () => {
    const { status } = await setup(Param)
    expect(status).toBe(Status.Success)
  })
})
