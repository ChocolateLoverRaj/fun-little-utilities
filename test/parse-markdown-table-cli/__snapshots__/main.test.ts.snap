// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`--format=dict --indentType=none calls console.info 1`] = `
Array [
  ActionInstance {
    "data": Array [
      "[{\\"id\\":\\"1\\",\\"name\\":\\"John Doe\\",\\"email\\":\\"john-doe@gmail.com\\"},{\\"id\\":\\"2\\",\\"name\\":\\"Peter Smith\\",\\"email\\":\\"petersmith22@outlook.com\\"},{\\"id\\":\\"3\\",\\"name\\":\\"Julia Jones\\",\\"email\\":\\"jjones778@gmail.com\\"}]",
    ],
    "type": "info",
  },
]
`;

exports[`--format=dict --indentType=space --indentSize=2 calls console.info 1`] = `
Array [
  ActionInstance {
    "data": Array [
      "[
  {
    \\"id\\": \\"1\\",
    \\"name\\": \\"John Doe\\",
    \\"email\\": \\"john-doe@gmail.com\\"
  },
  {
    \\"id\\": \\"2\\",
    \\"name\\": \\"Peter Smith\\",
    \\"email\\": \\"petersmith22@outlook.com\\"
  },
  {
    \\"id\\": \\"3\\",
    \\"name\\": \\"Julia Jones\\",
    \\"email\\": \\"jjones778@gmail.com\\"
  }
]",
    ],
    "type": "info",
  },
]
`;

exports[`--format=dict --indentType=tab calls console.info 1`] = `
Array [
  ActionInstance {
    "data": Array [
      "[
	{
		\\"id\\": \\"1\\",
		\\"name\\": \\"John Doe\\",
		\\"email\\": \\"john-doe@gmail.com\\"
	},
	{
		\\"id\\": \\"2\\",
		\\"name\\": \\"Peter Smith\\",
		\\"email\\": \\"petersmith22@outlook.com\\"
	},
	{
		\\"id\\": \\"3\\",
		\\"name\\": \\"Julia Jones\\",
		\\"email\\": \\"jjones778@gmail.com\\"
	}
]",
    ],
    "type": "info",
  },
]
`;

exports[`--format=list --indentType=space --indentSize=2 calls console.info 1`] = `
Array [
  ActionInstance {
    "data": Array [
      "{
  \\"headers\\": [
    \\"id\\",
    \\"name\\",
    \\"email\\"
  ],
  \\"rows\\": [
    [
      \\"1\\",
      \\"John Doe\\",
      \\"john-doe@gmail.com\\"
    ],
    [
      \\"2\\",
      \\"Peter Smith\\",
      \\"petersmith22@outlook.com\\"
    ],
    [
      \\"3\\",
      \\"Julia Jones\\",
      \\"jjones778@gmail.com\\"
    ]
  ]
}",
    ],
    "type": "info",
  },
]
`;

exports[`assertions InputParam yields multiple chunks 1`] = `
Array [
  "",
  "
",
  "  ",
  "| i",
  "d | ",
  "name ",
  "      ",
  " | emai",
  "l       ",
  "         ",
  "    |
  |-",
  "---|-------",
  "------|-----",
  "-------------",
  "--------|
  | ",
  " 1 | John Doe  ",
  "  | john-doe@gma",
  "il.com       |
  ",
  "|  2 | Peter Smith",
  " | petersmith22@out",
  "look.com |
  |  3 | ",
  "Julia Jones | jjones7",
  "78@gmail.com      |
",
]
`;
